File: 00-Code/Kotlin/Kotlin-CoroutineSample/src/main/java/com/bennyhuo/coroutines/sample/Ex10_JavaInterop.java
package com.bennyhuo.coroutines.sample;

import com.bennyhuo.coroutines.utils.LogKt;

/*在 java 中间接使用协程 */
public class Ex10_JavaInterop {
    public static void main(String... args) {
        try {
            LogKt.log(CoroutineInterop.loadString());
            LogKt.log(CoroutineInterop.loadString());
            LogKt.log(CoroutineInterop.loadFuture().get());
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: Kotlin/Code/Kotlin-CoroutineSample/src/main/java/com/bennyhuo/coroutines/sample/Ex10_JavaInterop.kt
@file:JvmName("CoroutineInterop")

package com.bennyhuo.coroutines.sample

import kotlinx.coroutines.experimental.async
import kotlinx.coroutines.experimental.delay
import kotlinx.coroutines.experimental.future.future
import kotlinx.coroutines.experimental.runBlocking

private suspend fun loadAsync(): String {
    delay(1000L)
    return "Hello"
}

fun loadString() = runBlocking {
    async { loadAsync() }.await()
}

//需要 org.jetbrains.kotlinx:kotlinx-coroutines-jdk8:0.22.5，Android 上面目前仍不方便使用
fun loadFuture() = future {
    async { loadAsync() }.await()
}